packages/repository/src/decorators/repository.ts

```diff
@@ -76,41 +84,79 @@ export class RepositoryMetadata {
 }
 
 /**
- * Decorator for model definitions
- * @param model Name of the repo or name/class of the model
- * @param dataSource Name or instance of the data source
- * @returns {(target:AnyType)}
+ * Decorator for repository injections on properties or method arguments
  *
- * For example:
+ * ```ts
+ * class CustomerController {
+ *   @repository(CustomerRepository.name) public custRepo: CustomerRepository;
  *
- * - @repository('myCustomerRepo')
- * - @repository('Customer', 'mysqlDataSource')
- * - @repository(Customer, mysqlDataSource)
- * - @repository('Customer', mysqlDataSource)
- * - @repository(Customer, 'mysqlDataSource')
+ *   constructor(
+ *     @repository(ProductRepository.name) public prodRepo: ProductRepository,
+ *   ) {}
+ *   // ...
+ * }
+ * ```
+ *
+ * @param repositoryName Name of the repo
  */
-export function repository<T extends Model>(
+export function repository(repositoryName: string): RepositoryDecorator;
+
+/**
+ * Decorator for DefaultCrudRepository generation and injection on properties
+ * or method arguments based on the given model and dataSource (or their names)
+ *
+ * ```ts
+ * class CustomerController {
+ *   @repository('Customer', 'mySqlDataSource')
+ *   public custRepo: DefaultCrudRepository<
+ *     Customer,
+ *     typeof Customer.prototype.id
+ *   >;
+ *
+ *   constructor(
+ *     @repository(Product, mySqlDataSource)
+ *     public prodRepo: DefaultCrudRepository<
+ *       Product,
+ *       typeof Product.prototype.id
+ *     >,
+ *   ) {}
+ *   // ...
+ * }
+ * ```
+ *
+ * @param model Name/class of the model
+ * @param dataSource Name/instance of the dataSource
+ */
+export function repository(
   model: string | typeof Entity,
+  dataSource: string | juggler.DataSource,
+): RepositoryDecorator;
+
+export function repository(
+  modelOrRepo: string | typeof Entity,
   dataSource?: string | juggler.DataSource,
 ) {
-  const meta = new RepositoryMetadata(model, dataSource);
+  const meta = new RepositoryMetadata(modelOrRepo, dataSource);
   return function(
     target: Object,
     key?: symbol | string,
-    descriptor?: TypedPropertyDescriptor<Repository<T>> | number,
+    descriptorOrIndex?: TypedPropertyDescriptor<BoundValue> | number,
```

Comment:
Please use `any` instead of `BoundValue`. `BoundValue` represents something that's bound in a Context. Here, you are describing a property of some type, not a value bound via `ctx.bind()`.